src/mechanics/cell/Cell.h

```diff
@@ -1,89 +1,85 @@
+    DofVector<double> operator()(const VectorOperation& op) override
     {
-        DofVector<double> gradient;
-        CellData cellData(mElements);
-        for (int iIP = 0; iIP < mIntegrationType.GetNumIntegrationPoints(); ++iIP)
-        {
-            auto ipCoords = mIntegrationType.GetLocalIntegrationPointCoordinates(iIP);
-            auto ipWeight = mIntegrationType.GetIntegrationPointWeight(iIP);
-            Jacobian jacobian(mCoordinateElement.ExtractNodeValues(),
-                                    mCoordinateElement.GetDerivativeShapeFunctions(ipCoords));
-            CellIpData cellipData(mElements, jacobian, ipCoords);
-            gradient += mIntegrand[iIP].Gradient(cellData, cellipData) * jacobian.Det() * ipWeight;
-        }
-        return gradient;
+        return Integrate(op, DofVector<double>());
+    }
+
+    DofMatrix<double> operator()(const MatrixOperation& op) override
+    {
+        return Integrate(op, DofMatrix<double>());
     }
 
-    //! @brief builds the hessian0 matrix
-    DofMatrix<double> Hessian0() override
+    double operator()(const ScalarOperation& op) override
     {
-        DofMatrix<double> hessian0;
-        CellData cellData(mElements);
+        return Integrate(op, double{0});
+    }
+
+    void operator()(const VoidOperation& op) override
+    {
+        CellData cellData(mElement);
         for (int iIP = 0; iIP < mIntegrationType.GetNumIntegrationPoints(); ++iIP)
         {
             auto ipCoords = mIntegrationType.GetLocalIntegrationPointCoordinates(iIP);
-            auto ipWeight = mIntegrationType.GetIntegrationPointWeight(iIP);
-            Jacobian jacobian(mCoordinateElement.ExtractNodeValues(),
-                                    mCoordinateElement.GetDerivativeShapeFunctions(ipCoords));
-            CellIpData cellipData(mElements, jacobian, ipCoords);
-            hessian0 += mIntegrand[iIP].Hessian0(cellData, cellipData) * jacobian.Det() * ipWeight;
+            Jacobian jacobian(mElement.CoordinateElement().ExtractNodeValues(),
+                              mElement.CoordinateElement().GetDerivativeShapeFunctions(ipCoords));
+            CellIpData cellipData(mElement, jacobian, ipCoords);
+            op(mIntegrands[iIP], cellData, cellipData);
         }
-        return hessian0;
     }
 
-
     DofVector<int> DofNumbering() override
     {
         return DofVector<int>();
     }
 
-    //! @brief Extracts a vector (each IP) of vectors (several IPValues for the same integrion point) of IPValues
-    std::vector<std::vector<IPValue>> IPValues() override
+private:
+    //! @brief integrates various operations with various return types
+    //! @param op operation to perform
+    //! @param result result value. It is not clear how to properly initialize an arbitrary TResult to zero. Thus, the
+    //! user has to provide it with this argument.
+    template <typename TOperation, typename TReturn>
+    TReturn Integrate(TOperation&& op, TReturn result)
     {
```

Comment:
You pass a copy of `TReturn return` and copy it again when returning? You have probably explained it before but why don't you want to use a reference here?

There should also be an overloaded (or default argument) `Integrate(TOperation&& op)` function for the cases above imo.

Reply:
Are you concerned with performance? You assume that all those values are copied. This is not correct. The compiler is forced to perform a `move` if you return by value. And probably, he will optimize the pass-by-value as well somehow. Even if he does not, the value is supposed to be zero/empty, thus small.

Reply:
It is not he performance I am worried about but the way this functions is being used.

This
```
TReturn result(bla);
Integrate(op, result);
```

is much easier to understand than passing a temporary object. I agree that you save 1 line of code though...

```
TReturn result = Integrate(op, TReturn(bla));
```

Reply:
Save two lines :)
~~~
return Integrate(op, ReturnObj(...));
// vs
ReturnObj bla;
Integrate(op, bla);
return bla;
~~~

Reply:
Addressing your comment regarding a default argument. If you do not provide e.g. `DofVector()` as an argument, the template method is not able to deduce the type of `TReturn`. But I think that this still nicely eliminates code duplication.

Reply:
From my personal point of view, there should be no passing of a reference at all
```
ReturnObj bla(IntegrateReturnObj(op));
```
Even though this requires to change the name of the function to define the type of the return object, this seems to me a much clearer interface, since the parameter TReturn is neither an input nor an output.

Reply:
I am happy to see a working implementation without a second argument. I didnt manage to do that. Maybe some `decltype` magic. Currently the argument has two purposes:
- enable template deduction
- provide a _zero_ value. E.g. double is not initialized to zero by default.